plugins/micromega/zify.ml

```diff
@@ -66,35 +67,49 @@ let is_convertible env evd t1 t2 = Reductionops.(is_conv env evd t1 t2)
 let get_type_of env evd e =
   Tacred.cbv_beta env evd (Retyping.get_type_of env evd e)
 
+(* arguments are dealt with in a second step *)
+
 let rec find_option pred l =
   match l with
   | [] -> raise Not_found
   | e :: l -> ( match pred e with Some r -> r | None -> find_option pred l )
 
-(** [HConstr] is a map indexed by EConstr.t.
-    It should only be used using closed terms.
- *)
+module ConstrMap = struct
+  open Names.GlobRef
+
+  type 'a t = 'a list Map.t
+
+  let add evd h e m =
+    let r = fst (EConstr.destRef evd h) in
+    try Map.modify r (fun _ l -> e :: l) m with Not_found -> Map.add r [e] m
+
+  let empty = Map.empty
+
+  let find evd h m =
+    match Map.find (fst (EConstr.destRef evd h)) m with
+    | e :: _ -> e
+    | [] -> assert false
+
+  let find_all evd h m = Map.find (fst (EConstr.destRef evd h)) m
```

Comment:
destko should be turned into Not_found here instead of adding destko cases in various places imo